File: sdks/java/src/main/java/org/apache/usergrid/java/client/UsergridClient.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package org.apache.usergrid.java.client;

import org.apache.usergrid.java.client.UsergridEnums.*;
import org.apache.usergrid.java.client.auth.UsergridAppAuth;
import org.apache.usergrid.java.client.auth.UsergridAuth;
import org.apache.usergrid.java.client.auth.UsergridUserAuth;
import org.apache.usergrid.java.client.model.*;
import org.apache.usergrid.java.client.query.UsergridQuery;
import org.apache.usergrid.java.client.response.UsergridResponse;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.util.HashMap;
import java.util.List;
import java.util.Map;

@SuppressWarnings("unused")
public class UsergridClient {

    @NotNull public static String DEFAULT_BASE_URL = "https://api.usergrid.com";

    @NotNull private UsergridClientConfig config;
    @Nullable private UsergridUser currentUser = null;
    @Nullable private UsergridAuth tempAuth = null;

    @NotNull private final UsergridRequestManager requestManager;

    public UsergridClient(@NotNull final UsergridClientConfig config) {
        this.config = config;
        this.requestManager = new UsergridRequestManager(this);
    }

    public UsergridClient(@NotNull final String orgId, @NotNull final String appId) {
        this(new UsergridClientConfig(orgId, appId));
    }

    public UsergridClient(@NotNull final String orgId, @NotNull final String appId, @NotNull final String baseUrl) {
        this(new UsergridClientConfig(orgId, appId, baseUrl));
    }

    public UsergridClient(@NotNull final String orgId, @NotNull final String appId, @NotNull final String baseUrl, @NotNull final UsergridAuthMode authMode) {
        this(new UsergridClientConfig(orgId, appId, baseUrl, authMode));
    }

    @NotNull public UsergridClientConfig getConfig() { return this.config; }
    public void setConfig(@NotNull final UsergridClientConfig config) { this.config = config; }

    @NotNull public String getAppId() { return this.config.appId; }
    public void setAppId(@NotNull final String appId) { this.config.appId = appId; }

    @NotNull public String getOrgId() { return this.config.orgId; }
    public void setOrgId(@NotNull final String orgId) { this.config.orgId = orgId; }

    @NotNull public String getBaseUrl() { return this.config.baseUrl; }
    public void setBaseUrl(@NotNull final String baseUrl) { this.config.baseUrl = baseUrl; }

    @NotNull public String clientAppUrl() { return getBaseUrl() + "/" + getOrgId() + "/" + getAppId(); }

    @NotNull public String managementUrl() { return getBaseUrl() + "/management"; }

    @NotNull public UsergridAuthMode getAuthMode() { return this.config.authMode; }
    public void setAuthMode(@NotNull final UsergridAuthMode authMode) { this.config.authMode = authMode; }

    @Nullable public UsergridUser getCurrentUser() { return this.currentUser; }
    public void setCurrentUser(@Nullable final UsergridUser currentUser) { this.currentUser = currentUser; }

    @Nullable public UsergridUserAuth getUserAuth() { return (this.currentUser != null) ? this.currentUser.getUserAuth() : null; }

    @Nullable public UsergridAppAuth getAppAuth() { return this.config.appAuth; }
    public void setAppAuth(@Nullable final UsergridAppAuth appAuth) { this.config.appAuth = appAuth; }

    @Nullable
    public UsergridAuth authForRequests() {
        UsergridAuth authForRequests = null;
        if (tempAuth != null) {
            if (tempAuth.isValidToken()) {
                authForRequests = tempAuth;
            }
            tempAuth = null;
        } else {
            switch (config.authMode) {
                case USER: {
                    if (this.currentUser != null && this.currentUser.getUserAuth() != null && this.currentUser.getUserAuth().isValidToken()) {
                        authForRequests = this.currentUser.getUserAuth();
                    }
                    break;
                }
                case APP: {
                    if (this.config.appAuth != null && this.config.appAuth.isValidToken()) {
                        authForRequests = this.config.appAuth;
                    }
                    break;
                }
            }
        }
        return authForRequests;
    }

    @NotNull
    public UsergridClient usingAuth(@Nullable final UsergridAuth auth) {
        this.tempAuth = auth;
        return this;
    }

    @NotNull
    public UsergridClient usingToken(@NotNull final String accessToken) {
        this.tempAuth = new UsergridAuth(accessToken);
        return this;
    }

    @NotNull
    public UsergridResponse authenticateApp() {
        if( this.config.appAuth == null ) {
            return UsergridResponse.fromError(this,  "Invalid UsergridAppAuth.", "UsergridClient's appAuth is null.");
        }
        return this.authenticateApp(this.config.appAuth);
    }

    @NotNull
    public UsergridResponse authenticateApp(@NotNull final UsergridAppAuth auth) {
        this.config.appAuth = auth;
        return this.requestManager.authenticateApp(auth);
    }

    @NotNull
    public UsergridResponse authenticateUser(@NotNull final UsergridUserAuth userAuth) {
        return this.authenticateUser(userAuth,true);
    }

    @NotNull
    public UsergridResponse authenticateUser(@NotNull final UsergridUserAuth userAuth, final boolean setAsCurrentUser) {
        UsergridResponse response = this.requestManager.authenticateUser(userAuth);
        if( response.ok() && setAsCurrentUser ) {
            this.setCurrentUser(response.user());
        }
        return response;
    }

    @NotNull
    public UsergridResponse resetPassword(@NotNull final UsergridUser user, @NotNull final String oldPassword, @NotNull final String newPassword) {
        String usernameOrEmail = user.usernameOrEmail();
        if( usernameOrEmail == null ) {
            return UsergridResponse.fromError(this,  "Error resetting password.", "The UsergridUser object must contain a valid username or email to reset the password.");
        }
        Map<String, Object> data = new HashMap<>();
        data.put("newpassword", newPassword);
        data.put("oldpassword", oldPassword);
        String[] pathSegments = { "users", usernameOrEmail, "password"};
        UsergridRequest request = new UsergridRequest(UsergridHttpMethod.POST, UsergridRequest.APPLICATION_JSON_MEDIA_TYPE, this.clientAppUrl(), null, data, this.authForRequests() ,pathSegments);
        return this.sendRequest(request);
    }

    @NotNull
    public UsergridResponse logoutCurrentUser()  {
        UsergridUser currentUser = this.currentUser;
        if( currentUser != null ) {
            String uuidOrUsername = currentUser.uuidOrUsername();
            UsergridUserAuth userAuth = currentUser.getUserAuth();
            if( uuidOrUsername != null && userAuth != null ) {
                String accessToken = userAuth.getAccessToken();
                if( accessToken != null ) {
                    return logoutUser(uuidOrUsername, accessToken);
                }
            }
        }
        return UsergridResponse.fromError(this,"UsergridClient's currentUser is not valid.", "UsergridClient's currentUser is null or has no uuid or username.");
    }

    @NotNull
    public UsergridResponse logoutUserAllTokens(@NotNull final String uuidOrUsername) {
        return logoutUser(uuidOrUsername, null);
    }

    @NotNull
    public UsergridResponse logoutUser(@NotNull final String uuidOrUsername, @Nullable final String token){
        String[] pathSegments = {"users", uuidOrUsername, ""};
        int len = pathSegments.length;
        Map<String, Object> param = new HashMap<>();
        if(token != null){
            pathSegments[len-1] = "revoketoken";
            param.put("token",token);
        }
        else{
            pathSegments[len-1] = "revoketokens";
        }
        UsergridRequest request = new UsergridRequest(UsergridHttpMethod.PUT, UsergridRequest.APPLICATION_JSON_MEDIA_TYPE, this.clientAppUrl(), param, null, this.authForRequests() , pathSegments);
        UsergridResponse response = this.sendRequest(request);
        UsergridUser currentUser = this.getCurrentUser();
        if( currentUser != null && response.ok() ) {
            if( uuidOrUsername.equalsIgnoreCase(currentUser.uuidOrUsername()) ) {
                this.setCurrentUser(null);
            }
        }
        return response;
    }

    @NotNull
    public UsergridResponse sendRequest(@NotNull final UsergridRequest request) {
        return this.requestManager.performRequest(request);
    }

    @NotNull
    public UsergridResponse GET(@NotNull final String type, @NotNull final String uuidOrName) {
        String[] pathSegments = {type, uuidOrName};
        UsergridRequest request = new UsergridRequest(UsergridHttpMethod.GET, UsergridRequest.APPLICATION_JSON_MEDIA_TYPE, this.clientAppUrl(), this.authForRequests() , pathSegments);
        return this.sendRequest(request);
    }

    @NotNull
    public UsergridResponse GET(@NotNull final String type) {
        String[] pathSegments = {type};
        UsergridRequest request = new UsergridRequest(UsergridHttpMethod.GET, UsergridRequest.APPLICATION_JSON_MEDIA_TYPE, this.clientAppUrl(), this.authForRequests() , pathSegments);
        return this.sendRequest(request);
    }

    @NotNull
    public UsergridResponse GET(@NotNull final UsergridQuery query) {
        String collectionName = query.getCollection();
        if( collectionName == null ) {
            return UsergridResponse.fromError(this,  "Query collection name missing.", "Query collection name is missing.");
        }
        UsergridRequest request = new UsergridRequest(UsergridHttpMethod.GET, UsergridRequest.APPLICATION_JSON_MEDIA_TYPE, this.clientAppUrl(), query, this.authForRequests() , collectionName);
        return this.sendRequest(request);
    }

    @NotNull
    public UsergridResponse PUT(@NotNull final String type, @NotNull final String uuidOrName, @NotNull final Map<String, ?> jsonBody) {
        String[] pathSegments = { type, uuidOrName };
        UsergridRequest request = new UsergridRequest(UsergridHttpMethod.PUT, UsergridRequest.APPLICATION_JSON_MEDIA_TYPE, this.clientAppUrl(), null, jsonBody, this.authForRequests() , pathSegments);
        return this.sendRequest(request);
    }

    @NotNull
    public UsergridResponse PUT(@NotNull final String type, @NotNull final Map<String, ?> jsonBody) {
        String uuidOrName = null;
        Object uuid = jsonBody.get(UsergridEntityProperties.UUID.toString());
        if( uuid != null ) {
            uuidOrName = uuid.toString();
        } else {
            Object name = jsonBody.get(UsergridEntityProperties.NAME.toString());
            if( name != null ) {
                uuidOrName = name.toString();
            }
        }
        if( uuidOrName == null ) {
            return UsergridResponse.fromError(this,  "jsonBody not valid..", "The `jsonBody` must contain a valid value for either `uuid` or `name`.");
        }
        String[] pathSegments = { type, uuidOrName };
        UsergridRequest request = new UsergridRequest(UsergridHttpMethod.PUT, UsergridRequest.APPLICATION_JSON_MEDIA_TYPE, this.clientAppUrl(), null, jsonBody, this.authForRequests() , pathSegments);
        return this.sendRequest(request);
    }

    @NotNull
    public UsergridResponse PUT(@NotNull final UsergridEntity entity) {
        String entityUuidOrName = entity.uuidOrName();
        if( entityUuidOrName == null ) {
            return UsergridResponse.fromError(this,  "No UUID or name found.", "The entity object must have a `uuid` or `name` assigned.");
        }
        String[] pathSegments = { entity.getType(), entityUuidOrName };
        UsergridRequest request = new UsergridRequest(UsergridHttpMethod.PUT, UsergridRequest.APPLICATION_JSON_MEDIA_TYPE, this.clientAppUrl(), null, entity, this.authForRequests() , pathSegments);
        return this.sendRequest(request);
    }

    @NotNull
    public UsergridResponse PUT(@NotNull final UsergridQuery query, @NotNull final Map<String, ?> jsonBody) {
        String collectionName = query.getCollection();
        if( collectionName == null ) {
            return UsergridResponse.fromError(this,  "Query collection name missing.", "Query collection name is missing.");
        }
        UsergridRequest request = new UsergridRequest(UsergridHttpMethod.PUT, UsergridRequest.APPLICATION_JSON_MEDIA_TYPE, this.clientAppUrl(), null, jsonBody, null, query, this.authForRequests(),collectionName);
        return this.sendRequest(request);
    }

    @NotNull
    public UsergridResponse POST(final @NotNull UsergridEntity entity) {
        String[] pathSegments = {entity.getType()};
        UsergridRequest request = new UsergridRequest(UsergridHttpMethod.POST, UsergridRequest.APPLICATION_JSON_MEDIA_TYPE, this.clientAppUrl(), null, entity, this.authForRequests() , pathSegments);
        return this.sendRequest(request);
    }

    @NotNull
    public UsergridResponse POST(@NotNull final List<UsergridEntity> entities) {
        if( entities.isEmpty() ) {
            return UsergridResponse.fromError(this,  "Unable to POST entities.", "entities array is empty.");
        }
        String[] pathSegments = {entities.get(0).getType()};
        UsergridRequest request = new UsergridRequest(UsergridHttpMethod.POST, UsergridRequest.APPLICATION_JSON_MEDIA_TYPE, this.clientAppUrl(), null, entities, this.authForRequests() , pathSegments);
        return this.sendRequest(request);
    }

    @NotNull
    public UsergridResponse POST(@NotNull final String type, @NotNull final String uuidOrName, @NotNull final Map<String, ?> jsonBody) {
        String[] pathSegments = {type, uuidOrName};
        UsergridRequest request = new UsergridRequest(UsergridHttpMethod.POST, UsergridRequest.APPLICATION_JSON_MEDIA_TYPE, this.clientAppUrl(), null, jsonBody, this.authForRequests() , pathSegments);
        return this.sendRequest(request);
    }

    @NotNull
    public UsergridResponse POST(@NotNull final String type, @NotNull final Map<String, ?> jsonBody) {
        UsergridRequest request = new UsergridRequest(UsergridHttpMethod.POST, UsergridRequest.APPLICATION_JSON_MEDIA_TYPE, this.clientAppUrl(), null, jsonBody, this.authForRequests() , type);
        return this.sendRequest(request);
    }

    @NotNull
    public UsergridResponse POST(@NotNull final String type, @NotNull final List<Map<String, ?>> jsonBodies) {
        UsergridRequest request = new UsergridRequest(UsergridHttpMethod.POST, UsergridRequest.APPLICATION_JSON_MEDIA_TYPE, this.clientAppUrl(), null, jsonBodies, this.authForRequests() , type);
        return this.sendRequest(request);
    }

    @NotNull
    public UsergridResponse DELETE(@NotNull final UsergridEntity entity) {
        String entityUuidOrName = entity.uuidOrName();
        if( entityUuidOrName == null ) {
            return UsergridResponse.fromError(this,  "No UUID or name found.", "The entity object must have a `uuid` or `name` assigned.");
        }
        String[] pathSegments = {entity.getType(), entityUuidOrName};
        UsergridRequest request = new UsergridRequest(UsergridHttpMethod.DELETE, UsergridRequest.APPLICATION_JSON_MEDIA_TYPE, this.clientAppUrl(), this.authForRequests() , pathSegments);
        return this.sendRequest(request);
    }

    @NotNull
    public UsergridResponse DELETE(@NotNull final String type, @NotNull final String uuidOrName) {
        String[] pathSegments = {type, uuidOrName};
        UsergridRequest request = new UsergridRequest(UsergridHttpMethod.DELETE, UsergridRequest.APPLICATION_JSON_MEDIA_TYPE, this.clientAppUrl(), this.authForRequests() , pathSegments);
        return this.sendRequest(request);
    }

    @NotNull
    public UsergridResponse DELETE(@NotNull final UsergridQuery query) {
        String collectionName = query.getCollection();
        if( collectionName == null ) {
            return UsergridResponse.fromError(this,  "Query collection name missing.", "Query collection name is missing.");
        }
        UsergridRequest request = new UsergridRequest(UsergridHttpMethod.DELETE, UsergridRequest.APPLICATION_JSON_MEDIA_TYPE, this.clientAppUrl(), query, this.authForRequests() , collectionName);
        return this.sendRequest(request);
    }

    @NotNull
    public UsergridResponse connect(@NotNull final UsergridEntity entity, @NotNull final String relationship, @NotNull final UsergridEntity to) {
        String entityUuidOrName = entity.uuidOrName();
        String toUuidOrName = to.uuidOrName();
        if( entityUuidOrName == null || toUuidOrName == null ) {
            return UsergridResponse.fromError(this, "Invalid Entity Connection Attempt.", "One or both entities that are attempting to be connected do not contain a valid UUID or Name property.");
        }
        return this.connect(entity.getType(), entityUuidOrName, relationship, to.getType(), toUuidOrName);
    }

    @NotNull
    public UsergridResponse connect(@NotNull final String entityType, @NotNull final String entityId, @NotNull final String relationship, @NotNull final String toType, @NotNull final String toName) {
        String[] pathSegments = {entityType, entityId, relationship, toType, toName};
        UsergridRequest request = new UsergridRequest(UsergridHttpMethod.POST, UsergridRequest.APPLICATION_JSON_MEDIA_TYPE, this.clientAppUrl(), this.authForRequests() , pathSegments);
        return this.sendRequest(request);
    }

    @NotNull
    public UsergridResponse connect(@NotNull final String entityType, @NotNull final String entityId, @NotNull final String relationship, @NotNull final String toId) {
        String[] pathSegments = { entityType, entityId, relationship, toId};
        UsergridRequest request = new UsergridRequest(UsergridHttpMethod.POST, UsergridRequest.APPLICATION_JSON_MEDIA_TYPE, this.clientAppUrl(), this.authForRequests() , pathSegments);
        return this.sendRequest(request);
    }

    @NotNull
    public UsergridResponse disconnect(@NotNull final String entityType, @NotNull final String entityId, @NotNull final String relationship, @NotNull final String fromUuid) {
        String[] pathSegments = {entityType, entityId, relationship, fromUuid};
        UsergridRequest request = new UsergridRequest(UsergridHttpMethod.DELETE, UsergridRequest.APPLICATION_JSON_MEDIA_TYPE, this.clientAppUrl(), this.authForRequests() , pathSegments);
        return this.sendRequest(request);
    }

    @NotNull
    public UsergridResponse disconnect(@NotNull final String entityType, @NotNull final String entityId, @NotNull final String relationship, @NotNull final String fromType, @NotNull final String fromName) {
        String[] pathSegments = {entityType, entityId, relationship, fromType, fromName};
        UsergridRequest request = new UsergridRequest(UsergridHttpMethod.DELETE, UsergridRequest.APPLICATION_JSON_MEDIA_TYPE, this.clientAppUrl(), this.authForRequests() , pathSegments);
        return this.sendRequest(request);
    }

    @NotNull
    public UsergridResponse disconnect(@NotNull final UsergridEntity entity, @NotNull final String relationship, @NotNull final UsergridEntity from) {
        String entityUuidOrName = entity.uuidOrName();
        String fromUuidOrName = from.uuidOrName();
        if( entityUuidOrName == null || fromUuidOrName == null ) {
            return UsergridResponse.fromError(this, "Invalid Entity Disconnect Attempt.", "One or both entities that are attempting to be disconnected do not contain a valid UUID or Name property.");
        }
        return this.disconnect(entity.getType(), entityUuidOrName, relationship, from.getType(), fromUuidOrName);
    }

    @NotNull
    public UsergridResponse getConnections(@NotNull final UsergridDirection direction, @NotNull final UsergridEntity entity, @NotNull final String relationship) {
        return this.getConnections(direction,entity,relationship,null);
    }

    @NotNull
    public UsergridResponse getConnections(@NotNull final UsergridDirection direction, @NotNull final UsergridEntity entity, @NotNull final String relationship, @Nullable final UsergridQuery query) {
        String entityUuidOrName = entity.uuidOrName();
        if( entityUuidOrName == null ) {
            return UsergridResponse.fromError(this, "Invalid Entity Get Connections Attempt.", "The entity must have a `uuid` or `name` assigned.");
        }
        return this.getConnections(direction,entity.getType(),entityUuidOrName,relationship,query);
    }

    @NotNull
    public UsergridResponse getConnections(@NotNull final UsergridDirection direction, @NotNull final String type, @NotNull final String uuidOrName, @NotNull final String relationship, @Nullable final UsergridQuery query) {
        String[] pathSegments = {type, uuidOrName, direction.connectionValue(), relationship};
        UsergridRequest request = new UsergridRequest(UsergridHttpMethod.GET, UsergridRequest.APPLICATION_JSON_MEDIA_TYPE, this.clientAppUrl(), query, this.authForRequests() , pathSegments);
        return this.sendRequest(request);
    }

    @NotNull
    public UsergridResponse getConnections(@NotNull final UsergridDirection direction, @NotNull final String uuid, @NotNull final String relationship, @Nullable final UsergridQuery query) {
        String[] pathSegments = {uuid, direction.connectionValue(), relationship};
        UsergridRequest request = new UsergridRequest(UsergridHttpMethod.GET, UsergridRequest.APPLICATION_JSON_MEDIA_TYPE, this.clientAppUrl(), query, this.authForRequests() , pathSegments);
        return this.sendRequest(request);
    }
}

File: sdks/python/usergrid/UsergridClient.py
# */
# * Licensed to the Apache Software Foundation (ASF) under one
# * or more contributor license agreements.  See the NOTICE file
# * distributed with this work for additional information
# * regarding copyright ownership.  The ASF licenses this file
# * to you under the Apache License, Version 2.0 (the
# * "License"); you may not use this file except in compliance
# * with the License.  You may obtain a copy of the License at
# *
# *   http://www.apache.org/licenses/LICENSE-2.0
# *
# * Unless required by applicable law or agreed to in writing,
# * software distributed under the License is distributed on an
# * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
# * KIND, either express or implied.  See the License for the
#    * specific language governing permissions and limitations
# * under the License.
# */

import json
import logging
import requests
from usergrid.UsergridAuth import UsergridAppAuth
from usergrid.app_templates import get_entity_url_template, post_collection_url_template, put_entity_url_template, \
    delete_entity_url_template, connect_entities_by_type_template, assign_role_url_template

__author__ = 'Jeff.West@yahoo.com'


def value_error(message):
    raise ValueError(message)


def usergrid_error(r):
    pass


class Usergrid(object):
    client = None

    @staticmethod
    def init(org_id,
             app_id,
             **kwargs):
        Usergrid.client = UsergridClient(org_id, app_id, **kwargs)

    @staticmethod
    def GET(collection, uuid_name, **kwargs):
        return Usergrid.client.GET(collection, uuid_name, **kwargs)

    @staticmethod
    def PUT(collection, uuid_name, data, **kwargs):
        return Usergrid.client.PUT(collection, uuid_name, data, **kwargs)

    @staticmethod
    def POST(collection, data, **kwargs):
        return Usergrid.client.POST(collection, data, **kwargs)

    @staticmethod
    def DELETE(collection, uuid_name, **kwargs):
        return Usergrid.client.DELETE(collection, uuid_name, **kwargs)

    @staticmethod
    def connect_entities(from_entity, relationship, to_entity, **kwargs):
        return Usergrid.client.connect_entities(from_entity, relationship, to_entity, **kwargs)

    @staticmethod
    def disconnect_entities(from_entity, relationship, to_entity, **kwargs):
        return Usergrid.client.disconnect_entities(from_entity, relationship, to_entity, **kwargs)

    @staticmethod
    def assign_role(role_uuid_name, user_entity, **kwargs):
        return Usergrid.client.assign_role(role_uuid_name, user_entity, **kwargs)


class UsergridResponse(object):
    def __init__(self, api_response, client):
        self.api_response = api_response
        self.client = client

        if api_response is None:
            self.ok = False
            self.body = 'No Response'

        else:
            self.headers = api_response.headers

            if api_response.status_code == 200:
                self.ok = True
                self.body = api_response.json()
                self.entities = self.body.get('entities', [])

            else:
                self.ok = False

                if api_response.headers.get('Content-type') == 'application/json':
                    self.body = api_response.json()
                else:
                    self.body = 'HTTP %s: %s' % (api_response.status_code, api_response.text)

    def __str__(self):
        return json.dumps(self.body)

    def first(self):
        return UsergridEntity(entity_data=self.entities[0]) if self.ok and self.entities and len(
                self.entities) > 0 else None

    def entity(self):
        return self.first()

    def last(self):
        return UsergridEntity(entity_data=self.entities[len(self.entities) - 1]) if self.ok and self.entities and len(
                self.entities) > 0 else None

    def has_next_page(self):
        return 'cursor' in self.body if self.ok else False


class UsergridEntity(object):
    def __init__(self, entity_data):
        self.entity_data = entity_data

    def __str__(self):
        return json.dumps(self.entity_data)

    def get(self, name, default=None):
        return self.entity_data.get(name, default)

    def entity_id(self):

        if self.entity_data.get('type', '').lower() in ['users', 'user']:
            return self.entity_data.get('uuid', self.entity_data.get('username'))

        return self.entity_data.get('uuid', self.entity_data.get('name'))

    def can_mutate_or_load(self):
        entity_id = self.entity_id()

        if entity_id is None or self.entity_data.get('type') is None:
            return False

        return True

    def put_property(self, name, value):
        self.entity_data[name] = value

    def put_properties(self, properties):
        if isinstance(properties, dict):
            self.entity_data.update(properties)

    def remove_property(self, name):

        if name is not None and name in self.entity_data:
            del self.entity_data[name]

    def remove_properties(self, properties):
        if isinstance(properties, (list, dict)):
            for property_name in properties:
                self.remove_property(property_name)

    def append(self, array_name, value):
        if array_name in self.entity_data:
            if isinstance(self.entity_data[array_name], list):
                self.entity_data[array_name].append(value)
        else:
            self.entity_data[array_name] = [value]

    def prepend(self, array_name, value):
        if array_name in self.entity_data:
            if isinstance(self.entity_data[array_name], list):
                self.entity_data[array_name].pre(value)
        else:
            self.entity_data[array_name] = [value]

    def insert(self, array_name, value, index):
        if array_name in self.entity_data:
            if isinstance(self.entity_data[array_name], list):
                self.entity_data[array_name].insert(index, value)

    def shift(self, array_name):
        if array_name in self.entity_data:
            if isinstance(self.entity_data[array_name], list):
                value = self.entity_data[array_name][0]
                self.entity_data[array_name] = self.entity_data[array_name][1:]
                return value

        return None

    def reload(self):
        if not self.can_mutate_or_load():
            raise ValueError('Unable to reload entity: No uuid nor name')

        response = Usergrid.GET(collection=self.entity_data.get('type'),
                                uuid_name=self.entity_id())
        if response.ok:
            self.entity_data.update(response.entity().entity_data)

        else:
            raise ValueError('Unable to reload entity: %s' % response)

    def save(self):
        if not self.can_mutate_or_load():
            raise ValueError('Unable to save entity: No uuid nor name')

        response = Usergrid.PUT(collection=self.entity_data.get('type'),
                                uuid_name=self.entity_id(),
                                data=self.entity_data)

        if response.ok and 'uuid' not in self.entity_data:
            self.entity_data['uuid'] = response.entity().get('uuid')

        return response

    def remove(self):
        if not self.can_mutate_or_load():
            raise ValueError('Unable to delete entity: No uuid nor name')

        return Usergrid.DELETE(collection=self.entity_data.get('type'),
                               uuid_name=self.entity_id())

    def get_connections(self, relationship, direction='connecting'):
        pass

    def connect(self, relationship, to_entity):

        if not to_entity.can_mutate_or_load():
            raise ValueError('Unable to connect to entity - no uuid or name')

        if not self.can_mutate_or_load():
            raise ValueError('Unable from connect to entity - no uuid or name')

        return Usergrid.connect_entities(self, relationship, to_entity)

    def disconnect(self, relationship, to_entity):
        if not to_entity.can_mutate_or_load():
            raise ValueError('Unable to connect to entity - no uuid or name')

        if not self.can_mutate_or_load():
            raise ValueError('Unable from connect to entity - no uuid or name')

        return Usergrid.disconnect_entities(self, relationship, to_entity)

    def attach_asset(self, filename, data, content_type):
        pass

    def download_asset(self, content_type=None):
        pass


class UsergridClient(object):
    def __init__(self,
                 org_id,
                 app_id,
                 base_url='http://api.usergrid.com',
                 client_id=None,
                 client_secret=None,
                 token_ttl_seconds=86400,
                 auth_fallback="none"):

        self.base_url = base_url
        self.org_id = org_id
        self.app_id = app_id
        self.auth_fallback = auth_fallback
        self.logger = logging.getLogger('usergrid.UsergridClient')
        self.session = requests.Session()

        self.url_data = {
            'base_url': base_url,
            'org_id': org_id,
            'app_id': app_id
        }

        if client_id and not client_secret:
            value_error('Client ID Specified but not Secret')

        elif client_secret and not client_id:
            value_error('Client ID Specified but not Secret')

        elif client_secret and client_id:
            self.auth = UsergridAppAuth(client_id=client_id,
                                        client_secret=client_secret,
                                        token_ttl_seconds=token_ttl_seconds)

            self.auth.authenticate(self)
            self.session.headers.update({'Authorization': 'Bearer %s' % self.auth.access_token})

    def __str__(self):
        return json.dumps({
            'base_url': self.base_url,
            'org_id': self.org_id,
            'app_id': self.app_id,
            'access_token': self.auth.access_token
        })

    def GET(self, collection, uuid_name, connections='none', auth=None, **kwargs):
        url = get_entity_url_template.format(collection=collection,
                                             uuid_name=uuid_name,
                                             connections=connections,
                                             **self.url_data)
        if auth:
            r = requests.get(url, headers={'Authorization': 'Bearer %s' % auth.access_token})

        else:
            r = self.session.get(url)

        return UsergridResponse(r, self)

    def PUT(self, collection, uuid_name, data, auth=None, **kwargs):
        url = put_entity_url_template.format(collection=collection,
                                             uuid_name=uuid_name,
                                             **self.url_data)

        if auth:
            r = requests.put(url,
                             data=json.dumps(data),
                             headers={'Authorization': 'Bearer %s' % auth.access_token})
        else:
            r = self.session.put(url, data=json.dumps(data))

        return UsergridResponse(r, self)

    def POST(self, collection, data, auth=None, **kwargs):
        url = post_collection_url_template.format(collection=collection,
                                                  **self.url_data)

        if auth:
            r = requests.post(url,
                              data=json.dumps(data),
                              headers={'Authorization': 'Bearer %s' % auth.access_token})
        else:
            r = self.session.post(url, data=json.dumps(data))

        return UsergridResponse(r, self)

    def DELETE(self, collection, uuid_name, auth=None, **kwargs):
        url = delete_entity_url_template.format(collection=collection,
                                                uuid_name=uuid_name,
                                                **self.url_data)

        if auth:
            r = requests.delete(url, headers={'Authorization': 'Bearer %s' % auth.access_token})
        else:
            r = self.session.delete(url)

        return UsergridResponse(r, self)

    def connect_entities(self, from_entity, relationship, to_entity, auth=None, **kwargs):

        url = connect_entities_by_type_template.format(from_collection=from_entity.get('type'),
                                                       from_uuid_name=from_entity.entity_id(),
                                                       relationship=relationship,
                                                       to_collection=to_entity.get('type'),
                                                       to_uuid_name=to_entity.entity_id(),
                                                       **self.url_data)

        if auth:
            r = requests.post(url, headers={'Authorization': 'Bearer %s' % auth.access_token})
        else:
            r = self.session.post(url)

        return UsergridResponse(r, self)

    def assign_role(self, role_uuid_name, entity, auth=None, **kwargs):
        url = assign_role_url_template.format(role_uuid_name=role_uuid_name,
                                              entity_type=entity.get('type'),
                                              entity_uuid_name=entity.entity_id(),
                                              **self.url_data)

        if auth:
            r = requests.delete(url, headers={'Authorization': 'Bearer %s' % auth.access_token})
        else:
            r = self.session.delete(url)

        return UsergridResponse(r, self)

    def disconnect_entities(self, from_entity, relationship, to_entity, auth=None, **kwargs):
            url = connect_entities_by_type_template.format(from_collection=from_entity.get('type'),
                                                           from_uuid_name=from_entity.entity_id(),
                                                           relationship=relationship,
                                                           to_collection=to_entity.get('type'),
                                                           to_uuid_name=to_entity.entity_id(),
                                                           **self.url_data)

            if auth:
                r = requests.delete(url, headers={'Authorization': 'Bearer %s' % auth.access_token})
            else:
                r = self.session.delete(url)

            return UsergridResponse(r, self)


class UsergridUser(object):
    def __init__(self):
        pass


class UsergridAsset(object):
    def __init__(self, filename, data, content_type):
        self.filename = filename
        self.data = data
        self.content_type = content_type
